phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -682,6 +687,49 @@ private void generateMutations(final TableRef tableRef, final long mutationTimes
         values.putAll(modifiedValues);
     }
 
+    private void annotateMutationsWithMetadata(PTable table, List<Mutation> rowMutations) {
+        //only annotate if the change detection flag is on the table and HBase supports
+        // preWALAppend coprocs server-side
+        if (table == null || !table.isChangeDetectionEnabled()
+            || !HbaseCompatCapabilities.hasPreWALAppend()) {
+            return;
+        }
+        //annotate each mutation with enough metadata so that anyone interested can
+        // deterministically figure out exactly what Phoenix schema object created the mutation
+        // Server-side we can annotate the HBase WAL with these.
+        for (Mutation mutation : rowMutations) {
+            annotateMutationWithMetadata(table, mutation);
+        }
+
+    }
+
+    public static void annotateMutationWithMetadata(PTable table, Mutation mutation) {
```

Comment:
Actually why not use the this method in the Ungroupped..Scanner in that for loop that goes over mutations? Also mutation  durability was not checked there I think

Reply:
Consolidated in WALAnnotationUtil as @ChinmaySKulkarni suggested.